c10/core/TensorImpl.cpp

```diff
@@ -225,25 +227,27 @@ at::DataPtr PlacementDeleteContext::makeDataPtr(
 AutogradMetaInterface::~AutogradMetaInterface() {}
 
 void TensorImpl::set_requires_grad(bool requires_grad) {
-  TORCH_INTERNAL_ASSERT(autograd_meta(), "set_requires_grad is not implemented for Tensor");
-  autograd_meta()->set_requires_grad(requires_grad, this);
+  TORCH_INTERNAL_ASSERT(type_set_.has(TensorTypeId::VariableTensorId), "set_requires_grad is not implemented for Tensor");
+  if (!requires_grad && !autograd_meta_) return;
+  if (!autograd_meta_) autograd_meta_ = impl::GetAutogradMetaFactory()->make();
+  autograd_meta_->set_requires_grad(requires_grad, this);
 }
 
 bool TensorImpl::requires_grad() const {
-  TORCH_INTERNAL_ASSERT(autograd_meta(), "requires_grad is not implemented for Tensor");
-  return autograd_meta()->requires_grad();
+  TORCH_INTERNAL_ASSERT(type_set_.has(TensorTypeId::VariableTensorId), "set_requires_grad is not implemented for Tensor");
+  if (!autograd_meta_) return false;
+  return autograd_meta_->requires_grad();
 }
 
 void TensorImpl::set_autograd_meta(std::unique_ptr<c10::AutogradMetaInterface> autograd_meta) {
+  // NB: autograd_meta may be null!  That just means it's the default
```

Comment:
Maybe add a comment here that reflects the discussion above about the autograd meta that can be in 3 possible states?